lib/SPIRV/SPIRVLowerBitCastToNonStandardType.cpp

```diff
@@ -0,0 +1,202 @@
+          }
+        }
+
+    for (auto &I : BCastsToNonStdVec)
+      Changed |= lowerBitCastToNonStdVec(I, InstsToErase);
+
+    while (!InstsToErase.empty()) {
+      InstsToErase.top()->eraseFromParent();
+      InstsToErase.pop();
+    }
+
+    return Changed ? PreservedAnalyses::none() : PreservedAnalyses::all();
+  }
+
+  bool lowerBitCastToNonStdVec(std::pair<Instruction *, VectorType *> &Inst,
+                               std::stack<Instruction *> &InstsToErase) {
+    bool Changed = false;
+    std::vector<Instruction *> InstsToInsert;
+    Instruction *I = Inst.first;
+    VectorType *DestVecTy = Inst.second;
+    uint64_t NumElemsInDestVec = DestVecTy->getElementCount().getValue();
+    InstsToErase.push(I);
+    AddrSpaceCastInst *Src = nullptr;
+    auto *ASCastInstIter = I;
+    while ((ASCastInstIter = ASCastInstIter->getNextNode())) {
+      if (auto *ASCastInst = dyn_cast<AddrSpaceCastInst>(ASCastInstIter)) {
+        unsigned DestAddrSpace = ASCastInst->getDestAddressSpace();
+        auto *SrcPointer = cast<CastInst>(I)->getSrcTy();
+        auto *SrcTy = cast<PointerType>(SrcPointer)->getElementType();
+        SrcPointer = SrcTy->getPointerTo(DestAddrSpace);
+        if (ASCastInst->getOperand(0) == cast<Value>(I)) {
+          Src = new AddrSpaceCastInst(I->getOperand(0), SrcPointer);
+          InstsToInsert.push_back(Src);
+          InstsToErase.push(ASCastInstIter);
+          break;
+        }
+      }
+    }
+    if (!ASCastInstIter)
+      ASCastInstIter = I;
+    auto *SrcTy =
+        cast<PointerType>(cast<Value>(Src)->getType())->getElementType();
+    auto *Load =
+        Src ? new LoadInst(SrcTy, Src, "", false, Align())
+            : new LoadInst(SrcTy, I->getOperand(0), "", false, Align());
+    InstsToInsert.push_back(Load);
+    auto *LoadInstIter = ASCastInstIter;
+    while ((LoadInstIter = LoadInstIter->getNextNode())) {
+      if (auto *LI = dyn_cast<LoadInst>(LoadInstIter))
+        if (LI->getOperand(0) == ASCastInstIter) {
+          InstsToErase.push(LoadInstIter);
+          break;
+        }
+    }
+    if (!LoadInstIter)
+      LoadInstIter = ASCastInstIter;
+    uint64_t NumElemsInSrcVec =
+        cast<VectorType>(SrcTy)->getElementCount().getValue();
+    int ElemIdx = 0;
+    Instruction *ExtrElInstIter = LoadInstIter;
+    while ((ASCastInstIter = ASCastInstIter->getNextNode())) {
+      if (auto *EEI = dyn_cast<ExtractElementInst>(ASCastInstIter)) {
+        if (EEI->getOperand(0) == cast<Value>(LoadInstIter)) {
+          ElemIdx = cast<ConstantInt>(EEI->getIndexOperand())->getSExtValue() /
+                    (NumElemsInDestVec / NumElemsInSrcVec);
+          ExtrElInstIter = ASCastInstIter;
+          InstsToErase.push(ExtrElInstIter);
+          break;
+        }
+      }
+    }
+    auto *ExtractElement = ExtractElementInst::Create(
+        Load, ConstantInt::get(Type::getInt64Ty(I->getContext()), ElemIdx));
+    InstsToInsert.push_back(ExtractElement);
```

Comment:
It is simpler to use llvm::IRBuilder<> instead of saving instructions just for setting an order of insertion.